### PR TITLE
Add Kconfigs for `PSA_WANT` symbols

### PR DESCRIPTION
**Pull request by valeriosetti**
With the goal to move toward PSA crypto APIs (https://github.com/zephyrproject-rtos/zephyr/issues/43712) it is useful to have the possibility to define `PSA_WANT_xxx` features that are required. So far all of these symbols are automatically guessed in MbedTLS' header files by starting from `MBEDTLS_xxx` features. However there are some limitation to this:

- `PSA_WANT_xxx` symbols are not 1:1 mapped with `MBEDTLS_xxx` ones
- for some `PSA_WANT_xxx` there is no mapping with `MBEDTLS_xxx` (ex: `PSA_WANT_KEY_TYPE_DERIVE`)

For this reason this PR introduces `Kconfig` and an header file to support `PSA_WANT_xxx` symbols with a mechanism which mimics the already existing `Kconfig.tls-generic`/`config-tls-generic.h` combo. These new Kconfigs will be used whenever there is some crypto provider enabled, i.e. when `MBEDTLS_PSA_CRYPTO_C` or `BUILD_WITH_TFM` are set.

### CODE REVIEW

**Comment by valeriosetti** [comment]
CI is green so this PR is ready for initial reviews.

File: modules/mbedtls/Kconfig.tls-generic
Comment (frkv) [inline modules/mbedtls/Kconfig.tls-generic:491]: This should be MBEDTLS_PSA_CRYPTO_CLIENT

Comment (valeriosetti) [reply modules/mbedtls/Kconfig.tls-generic]: Correct, thanks! Fixed

**Review by ithinuel — COMMENTED** [review]
That is a lot of new config. I didn't check each individually but it looks alright.

IIUC `Kconfig.psa`and `config-psa.h` are generated by  `create_psa_files.py`.

- Is it expected to be executed manually every now and then?
- Is there any CI setup that can run it an check it does not generate any change?
  It would be nice to have this to detect changes in compatibility/naming.

File: tests/bsim/bluetooth/mesh/overlay_psa.conf
Comment (alxelax) [inline tests/bsim/bluetooth/mesh/overlay_psa.conf:8]: I do not think I agree that BLE Mesh requires all crypto PSA features. There is list of features that Mesh uses, those are `selected` during `BT_MESH_USES_MBEDTLS_PSA` activation.

**Review by alxelax — DISMISSED** [review]

File: tests/bsim/bluetooth/mesh/overlay_psa.conf
Comment (alxelax) [reply tests/bsim/bluetooth/mesh/overlay_psa.conf:8]: There is one more thing. As a Bluetooth Mesh user, I expect that when I chose crypto backend, this is sufficient to build my application. This extra option seems like hidden knowledge that isn't obvious to people not known mbedtls-psa configuration system.

Comment (tomi-font) [reply tests/bsim/bluetooth/mesh/overlay_psa.conf:8]: > I do not think I agree that BLE Mesh requires all crypto PSA features. There is list of features that Mesh uses, those are `selected` during `BT_MESH_USES_MBEDTLS_PSA` activation.

So far the PSA configuration has been defined by MbedTLS's default configuration file, `psa/crypto_config.h`, which unconditionally does `#define PSA_WANT_[...] 1` for every symbol. In this regard this PR does not change anything. It introduces the configuration of `PSA_WANT` symbols for future work to build on top of this. Also note that when TF-M is in use, currently the actual crypto features available are defined by what TF-M profile is used.

However I am not certain how this fares when TF-M is not in use. It could be that with MbedTLS's adjustment of legacy symbols from PSA ones this would inflate the final code size for nothing.

> There is one more thing. As a Bluetooth Mesh user, I expect that when I chose crypto backend, this is sufficient to build my application. This extra option seems like hidden knowledge that isn't obvious to people not known mbedtls-psa configuration system.

Good point. Could we instead `select PSA_CRYPTO_ENABLE_ALL` @valeriosetti? (And you may also want to add it under `config BT_MESH_USES_TFM_PSA`, actually.) Or then, because of the first point above, select only needed PSA symbols. Or simply stick to the legacy ones and `MBEDTLS_USE_PSA_CRYPTO` if possible?

File: modules/mbedtls/create_psa_files.py
Comment (tomi-font) [inline modules/mbedtls/create_psa_files.py:30]:
```suggestion
	bool "All PSA features"
```

**Comment by tomi-font** [comment]
> Is it expected to be executed manually every now and then?

It would need to be run every time we 1. want to make a change to something in `Kconfig.psa` or 2. update MbedTLS.

> Is there any CI setup that can run it an check it does not generate any change?
It would be nice to have this to detect changes in compatibility/naming.

I have also thought about the possibility of having an automatic check to ensure that the committed header file is exactly what is generated by the script. Could be nice if it's not too involved to do.

File: tests/bsim/bluetooth/mesh/overlay_psa.conf
Comment (valeriosetti) [reply tests/bsim/bluetooth/mesh/overlay_psa.conf:8]: As @alxelax suggested `BT_MESH_USES_MBEDTLS_PSA` selects required MbedTLS features, so I can try to replace legacy `MBEDTLS_` symbols with proper `PSA_WANT_` ones. This should fix both comments.
I'll give this a try

Comment (valeriosetti) [reply tests/bsim/bluetooth/mesh/overlay_psa.conf]: @alxelax I reverted the previous change and I've added https://github.com/zephyrproject-rtos/zephyr/pull/72243/commits/513fb7d339a697eae7b7c1920123f0e909c927a9 which should match both your requirements. Wdyt?

File: modules/mbedtls/Kconfig.tls-generic
Comment (tomi-font) [inline modules/mbedtls/Kconfig.tls-generic:108]: Are we gonna have other MbedTLS Kconfig options start depending on PSA ones? :thinking:

File: subsys/bluetooth/mesh/Kconfig
Comment (tomi-font) [inline subsys/bluetooth/mesh/Kconfig:1354]: What line(s) is this comment addressing?

File: modules/mbedtls/create_psa_files.py
Comment (valeriosetti) [reply modules/mbedtls/create_psa_files.py]: All prompts and description for `PSA_CRYPTO_ENABLE_ALL` should be fixed now

**Comment by valeriosetti** [comment]
> > Is there any CI setup that can run it an check it does not generate any change?
> > It would be nice to have this to detect changes in compatibility/naming.
> 
> I have also thought about the possibility of having an automatic check to ensure that the committed header file is exactly what is generated by the script. Could be nice if it's not too involved to do.

I totally agree on this and I will investigate on it.

File: subsys/bluetooth/mesh/Kconfig
Comment (valeriosetti) [reply subsys/bluetooth/mesh/Kconfig:1354]: The line below. It enables RSA keys support which might look odd since we're only using ECDH and ECDSA, so I thought adding a comment there could help. Doesn't it?

Comment (tomi-font) [reply subsys/bluetooth/mesh/Kconfig:1354]: I see. I didn't pay much attention to what is (meant to be) used compared to what is enabled. Others can likely judge better. The comment is fine and makes sense.

File: modules/mbedtls/Kconfig.tls-generic
Comment (frkv) [reply modules/mbedtls/Kconfig.tls-generic:108]: Would it be more convenient (at least for the reader) to try to emulate the "CAN"-concept from Mbed TLS as Kconfigs?

Like `MBEDTLS_CAN_ECDH`
https://github.com/Mbed-TLS/mbedtls/blob/2ca6c285a0dd3f33982dd57299012dacab1ff206/include/mbedtls/config_adjust_legacy_crypto.h#L309C9-L309C25

Or more distinct: `MBEDTLS_PK_CAN_ECDSA_SOME`
https://github.com/Mbed-TLS/mbedtls/blob/2ca6c285a0dd3f33982dd57299012dacab1ff206/include/mbedtls/config_adjust_legacy_crypto.h#L333C9-L333C34

There are more than just `PSA_WANTS_ALG_XXXX` that can be used as a signal, since you can choose to use or not use PSA crypto with `MBEDTLS_USE_PSA_CRYPTO`. Arguably we of course prefer the `MBEDTLS_USE_PSA_CRYPTO` and PSA crypto will at some point likely be the only option, but it is still slightly more readable no not have to rehash both legacy and PSA configs for every ciphersuite

When I start thinking about it, there is an option to have a stable configuration follow the pattern `MBEDTLS_CAN_ECDH_ECDSA_KEY_EXCHANGE` for every relevant key exchange/cipher-suite

We have made a similar concept in nRF Connect SDK

Comment (valeriosetti) [reply modules/mbedtls/Kconfig.tls-generic:108]: TBH in my first attempt I was about to remove those "depends on" because this is totally duplicating checks that MbedTLS already performs in `check_config.h` and, in case something is missing, it already prints a clear message of what's missing and where to look at.
Honestly I'm not a huge fan of duplicating also the `MBEDTLS_CAN_ECDH` and other MbedTLS internal symbols, because this adds extra burden of work every time Zephyr updates the MbedTLS library because you need to update all these checks again and again.
In my opinion I would follow the same path I've introduced for the `PSA_WANT` symbols: just add 1 Kconfig symbol for every MbedTLS build symbol, but no extra logic. Then if something is not correct `check_config.h` will error out some message. No extra work for Zephyr. However this proposed approach imposes quite some change to the current `Kconfig.tls-generic" file so I would keep this for a follow up PR in case we find an agreement (@tomi-font ).
Wdyt?

Comment (tomi-font) [reply modules/mbedtls/Kconfig.tls-generic:108]: Agreed to keep further significant work for a follow-up PR. And at first thought it seems senseful to get rid of `depends on` statements if those checks are performed by MbedTLS itself.

**Comment by fabiobaltieri** [comment]
> I totally agree on this and I will investigate on it.

Since the script is in modules there's no current workflow that looks for pytest in there. I think you have two options:

- you could write a pytest test and add the path to https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/scripts_tests.yml#L71 (then you can probably mock the write function and make sure it's called with what you expect, you may have to refactor the script a bit to make it testable), or
- just add a test to validate against a golden sample in an actual build only test, see https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities for an example, this may be easier if you already have a build test that runs this script, just add the validation step on it or something

File: tests/bsim/bluetooth/mesh/overlay_psa.conf
Comment (alxelax) [reply tests/bsim/bluetooth/mesh/overlay_psa.conf]: Yep, now it looks much better. Seems it missed some option(s) since devices cannot generate public key in bsim tests.

Comment (valeriosetti) [reply tests/bsim/bluetooth/mesh/overlay_psa.conf]: Yes, I should have fixed both EC key generation issue and a missing  `PSA_WANT_ALG_ECB_NO_PADDING`. Hopefully bsim will be fine now.

**Review by tomi-font — DISMISSED** [review]
ping @ceolin @ithinuel @d3zd3z

File: modules/mbedtls/Kconfig.psa
Comment (valeriosetti) [reply modules/mbedtls/Kconfig.psa:5]: While implementing this I realized that the only relevant info to be added here is probably the file that is used as reference to generate the other ones. The reason for skipping version/revision is that MbedTLS can change (likely be upgraded over time), but what we really care of here is the `crypto_config.h` file. If that file does not change, then PSA is not introducing new symbols so Zephyr does not need to adapt.

Comment (valeriosetti) [reply modules/mbedtls/Kconfig.psa]: This should be fixed now

**Comment by valeriosetti** [comment]
> > I totally agree on this and I will investigate on it.
> 
> Since the script is in modules there's no current workflow that looks for pytest in there. I think you have two options:
> 
>     * you could write a pytest test and add the path to https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/scripts_tests.yml#L71 (then you can probably mock the write function and make sure it's called with what you expect, you may have to refactor the script a bit to make it testable), or
> 
>     * just add a test to validate against a golden sample in an actual build only test, see https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities for an example, this may be easier if you already have a build test that runs this script, just add the validation step on it or something

@fabiobaltieri I think I need a quick help on this. I modified the python script so that it's also able to just check if the already existing files are up to date or not and I also added a check in `twister.yaml`. However I don't see this test appearing on the CI. What am I missing?

**Review by tomi-font — DISMISSED** [review]

File: modules/mbedtls/Kconfig.psa
Comment (de-nordic) [reply modules/mbedtls/Kconfig.psa]: I guess ok.

**Review by ithinuel — DISMISSED** [review]

File: .github/workflows/twister.yaml
Comment (fabiobaltieri) [inline .github/workflows/twister.yaml:213]: This should not really be in twister.yaml, it could go in a different workflow but since it's implemented as a check option rather than an actualy pytest I don't there's a good place to put it.

How about instead of trying to run it directly from a workflow you add the check to one of the existing test or samples? (like the other example I pointed out in my comment). Maybe this could live as part of the tfm_integration test

Comment (valeriosetti) [reply .github/workflows/twister.yaml:213]: >  it could go in a different workflow

Initially I added this check in `scripts_test.yaml` as you suggested, but that workfile is executed on the Zephyr repo itself while my script requires the MbedTLS repo to be present as well in order to work properly (i.e. `west init` must be issued). That's why I moved it from `scripts_test.yaml` to `twister.yaml`

> How about instead of trying to run it directly from a workflow you add the check to one of the existing test or samples?

I will try this option

File: modules/mbedtls/Kconfig.tls-generic
Comment (frkv) [reply modules/mbedtls/Kconfig.tls-generic:108]: This is for sure follow-up work. This only makes sense if we have stable definitions in Mbed TLS distribution. Arguably PSA_WANT_ALG_XXXX should be sufficient for most cases, but then this can only happen when mbedtls_ prefixed crypto toolbox APIs are removed from the system (e.g. in downstream Zephyr or in upstream Mbed TLS)

File: modules/mbedtls/Kconfig.psa
Comment (frkv) [inline modules/mbedtls/Kconfig.psa:17]:
```suggestion
	bool "Enable all PSA crypto features"
```

Minor NIT

File: subsys/bluetooth/mesh/Kconfig
Comment (frkv) [reply subsys/bluetooth/mesh/Kconfig:1356]:
```suggestion
	# Required, otherwise there are missing symbols for PK helpers at link time.
```
I think a comma is needed

**Review by frkv — APPROVED** [review]
LGTM (minor NITS and assuming correct placement of the check-script is resolved)

File: .github/workflows/twister.yaml
Comment (valeriosetti) [reply .github/workflows/twister.yaml]: Turns out that I overlooked the 2nd option you wrote [previously](https://github.com/zephyrproject-rtos/zephyr/pull/72243#issuecomment-2111907143), but it was actually the easiest way to achieve what I was meant to do here. This should now be fixed in https://github.com/zephyrproject-rtos/zephyr/commit/37769f4f74fc5754a4e8f75abd9131eea68ed8cb.
Thanks again!

File: modules/mbedtls/Kconfig.psa
Comment (valeriosetti) [reply modules/mbedtls/Kconfig.psa:17]: I think I had issues where the CI was complaining about prompt starting with "Enable", so that's why I removed that part. I can add the "crypto" though

**Comment by valeriosetti** [comment]
I should have addressed all the comments now and the check on auto-generated files is active (implementation is [here](https://github.com/zephyrproject-rtos/zephyr/pull/72243/commits/22e7ac3922c2aef8b737af7e96cd3e50111038d1))

**Review by tomi-font — APPROVED** [review]
Very nice job!

Asking @fabiobaltieri as I couldn't find anything relating to the newly added check in the CI logs. Does it seem okay to you now, and will it be run in CI both if something in `modules/mbedtls` is modified and also when MbedTLS's manifest revision is changed?

**Review by ceolin — APPROVED** [review]

**Review by ithinuel — APPROVED** [review]

**Comment by valeriosetti** [comment]
Would it possible to merge this PR now that we have a good amount of approvals, please? This helps unlocking other depending PRs

**Comment by aescolar** [comment]
@valeriosetti Given that this can cause problems for some apps, should we have an entry in the migration guide?

**Comment by valeriosetti** [comment]
Yes sure! I can work on that. I was also requested to review the crypto documentation changes in https://github.com/zephyrproject-rtos/zephyr/pull/73241

**Pull request merged**
